### PR TITLE
Playwright framework with sample live chat widget test for chat button #1

### PR DESCRIPTION
**Pull request by gitmahadev**
Playwright framework with sample live chat widget test for chat button

For custom live chat widget, currently using sampleWidget from sample folder. as on required for test case will create configured script in automation_tests folder .

Adding all tests files under automation_tests

Run below command to execute automation tests,

- npm run test:build : to build sampleWidget and run tests on latest sampleWidget script.
- npm run test: to run tests on existing sampleWidget script.

Please refer test [as screenshot image]
169195474-c91e0458-3ccd-49f4-9649-f285a9dfc0b4

image

### CODE REVIEW

File: chat-widget/automation_tests/configuration/CustomEnvironment.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
+const PlaywrightEnvironment =
+  require("jest-playwright-preset/lib/PlaywrightEnvironment").default;
+/* eslint-disable @typescript-eslint/no-var-requires */
+const fs = require("fs");
+class CustomEnvironment extends PlaywrightEnvironment {
+    async handleTestEvent(event) {
+        if (event.name === "test_done" && event.test.errors.length > 0) {
```
Comment (xTEddie) [inline chat-widget/automation_tests/configuration/CustomEnvironment.js:8]: `test_done` as constant?

File: chat-widget/automation_tests/configuration/CustomEnvironment.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            try {
+                const parentName = event.test.parent.name.replace(/\W/g, "_");
+                const specName = event.test.name.replace(/\W/g, "_");
+                const contexts = this.global.browser.contexts();
+                for (let [contextIndex, currentContext] of contexts.entries()) {
+                    const contextPages = currentContext.pages();
+                    for (let [pageIndex, currentPage] of contextPages.entries()) {
+                        if (currentPage) {
+                            try {
+                                await currentPage.screenshot({
```
Comment (charliewang95) [inline chat-widget/automation_tests/configuration/CustomEnvironment.js:19]: Why are we taking screenshot here?

File: chat-widget/automation_tests/e2e/utility/constants.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+export enum CustomLiveChatWidgetConstants {
+  CustomLiveChatWidgetFilePath = "customlivechatwidgets/CustomLiveChatWidget.html",
+  LiveChatButtomId = "#oc-lcw-chat-button-title",
```
Comment (charliewang95) [inline chat-widget/automation_tests/e2e/utility/constants.ts:3]: "Button"

File: chat-widget/automation_tests/readme.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# How to run playwright testcases against your environment
```
Comment (charliewang95) [inline chat-widget/automation_tests/readme.md:1]: Please modify this file and remove unused document sections

File: chat-widget/automation_tests/package.json [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    "ts-jest": "^25.3.1",
+    "tslint-circular-dependencies": "^0.1.0",
+    "typescript": "^3.9.9"
+  },
+  "scripts": {
+    "compile": "tsc",
+    "clear_jest": "jest --clearCache",
+    "lint": "eslint .",
+    "test": "jest --detectOpenHandles --runInBand --passWithNoTests",
+    "test:build": "npm run --prefix ../../chat-widget build-sample && npm run test"
```
Comment (charliewang95) [inline chat-widget/automation_tests/package.json:29]: Same, use yarn here.

**Comment by gitmahadev** [comment]
Done

File: chat-widget/automation_tests/package-lock.json [hunk 6]
```diff
@@ -0,0 +1,19758 @@
...
+        "ts-jest": "^25.3.1",
+        "tslint-circular-dependencies": "^0.1.0",
+        "typescript": "^3.9.9"
+      }
+    },
+    "node_modules/@ampproject/remapping": {
+      "version": "2.2.0",
+      "resolved": "https://registry.npmjs.org/@ampproject/remapping/-/remapping-2.2.0.tgz",
+      "integrity": "sha512-qRmjj8nj9qmLTQXXmaR1cck3UXSRMPrbsLJAasZpF+t3riI71BXed5ebIOYwQntykeZuhjsdweEc9BxH5Jc26w==",
+      "dependencies": {
```
[the remark this answers is not in the document]
Comment (gitmahadev) [reply chat-widget/automation_tests/package-lock.json]: Done

File: chat-widget/automation_tests/readme.md (hunk 4 above)
Comment (gitmahadev) [reply chat-widget/automation_tests/readme.md]: Done

File: chat-widget/automation_tests/e2e/utility/constants.ts (hunk 3 above)
Comment (gitmahadev) [reply chat-widget/automation_tests/e2e/utility/constants.ts:3]: Done

File: chat-widget/automation_tests/configuration/CustomEnvironment.js (hunk 2 above)
Comment (gitmahadev) [reply chat-widget/automation_tests/configuration/CustomEnvironment.js:19]: it is used to capture screenshot if there is an error in testcase
as we are using playwright with jest then this is required.

I will check if we can use Playwright Test runner instead of Jest runner.

File: chat-widget/automation_tests/yarn.lock [hunk 7]
```diff
@@ -0,0 +1,6298 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (charliewang95) [inline chat-widget/automation_tests/yarn.lock:1]: Why are there 2 yarn.lock files modified here? Can we revert the changes for this one "/chat-widget/yarn.lock"?

Comment (gitmahadev) [reply chat-widget/automation_tests/yarn.lock:1]: Oops. seems yarn.lock modified while running yarn command. I will revert '/chat-widget/yarn.lock' file.

**Review by charliewang95 — APPROVED** [review]

**Comment by gitmahadev** [comment]
raised new PR. did not require this one now.
